### DISCUSSION TITLE
Custom Color Definition for Variants

### DISCUSSION THREAD

**Discussion by GreenInnovator** [GMOD/jbrowse-components#3757] [Q&A]
Hi, I'm trying to get get a custom color for variants based on an annotation field in the INFO of the vcf header, for example the AC or ANN attribute. Expressions I tried:
 `"color1": "jexl:get(feature, INFO.AC)==12?'green':'purple'"`  but that doesn't work as expected. 
`"color2": "jexl:get(feature.data,INFO.AC) == 12 ? 'green' : 'purple'"` throws an  "TypeErrer e.get is not a function" error.

I was able to successfully print the value of the INFO.AC or INFO.ANN fields to the console using the following expression:
` "color3": "jexl:log(feature.data,INFO.AC)"}`

The ultimate goal of this customization is to assign a custom color based on the Annotation_Impact generated by SNPeff, which is located in the third field of the INFO.ANN array, separated by pipes.
![image](https://github.com/GMOD/jbrowse-components/assets/18717239/729342e1-2508-42a0-8409-60d200292883)

Your guidance or expertise  would be appreciated.

Regards, Jan

**Comment by cmdcolin** [comment]
Note: you may want to use a JavaScript callback added by a plugin. writing sophisticated logic in jexl is tricky https://jbrowse.org/jb2/docs/config_guides/customizing_feature_colors/

I believe you'll want in jexl form `get(feature, 'INFO').AC[0]` or in JavaScript form `feature.get('INFO').AC[0]` (replace AC with ANN). not also that I refer to the zeroth element. Anything in INFO is an array (which is slightly annoying for single element things but reduces reasoning needed to check if something is an array or not...it's just always an array. also would avoid accessing feature.data directly, the get method generally accesses that for you

**Comment by GreenInnovator** [comment]
Thank you, the code modification worked and I was able to create a functional no-build plugin that is working as intended:
```
    install(pluginManager) {
      pluginManager.jexl.addFunction('customColor', feature => {
        const infoANN = feature.get('INFO').ANN;
        if (infoANN && infoANN.length > 0) {
          for (const ann of infoANN) {
            const annArray = ann.split('|');
            if (annArray.length >= 3) {
              const annotationImpact = annArray[2];
              if (annotationImpact === 'HIGH') {
                return 'red';
              } else if (annotationImpact === 'MODERATE') {
                return 'orange';
              } else if (annotationImpact === 'LOW') {
                return 'yellow';
              } else if (annotationImpact === 'MODIFIER') {
                return 'green';
              }
            }
          }
        }
      });
    }
```

**Reply by cmdcolin** [reply]
awesome! glad that works
